### ISSUE TITLE
MathJax hangs Chrome

### ISSUE THREAD

**Issue by ricardobeat** [mathjax/MathJax#548]
Hi,

It's about the third time this week I've got my browser completely frozen from opening a blog post. I noticed they all hang with the little "Typesetting Math: x%" at the bottom left, and found this library.

I'm running Chrome 28.0.1500.71 on OSX Mountain Lion. Here is one such post: http://www.leancrew.com/all-this/2013/08/hyperloop/

**Comment by pkra** [comment]
> It looks like a webkit bug. 

Google forked WebKit into Blink.

> Maybe there is something that can be tweaked in MathJax to avoid the bug or be comunicated to the Chromium team. 

Yes, that would be great. We could never reproduce the various issues described on this thread so we could neither pursue a fix on our end or submit a bug with with the Blink team.

**Comment by ManuelLlorens** [comment]
I have reduced the html file that hungs my Nexus 7 to this:
<pre><html><head><script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML.js"></script></head><body>\[a_1\]</body></html></pre>

I have traced the hung in MathJax code to this function (I have pretty printed the minified version of MathJax, so the names of variables will not match the original ones):

<pre>prepareOutput: function(c, f) {
                while (c.j < c.jaxIDs.length) {
                    var e = c.jaxIDs[c.j],
                        d = MathJax.OutputJax[e];
                    if (d[f]) {
                        //try {
                            var a = d[f](c);
                            if (typeof a === "function") {
                                if (a.called) {
                                    continue
                                }
                                this.RestartAfter(a)
                            }
                        /*} catch (b) {
                            if (!b.restart) {
                                MathJax.Message.Set(["PrepError", "Error preparing %1 output (%2)", e, f], null, 600);
                                MathJax.Hub.lastPrepError = b;
                                c.j++
                            }
                            return MathJax.Callback.After(["prepareOutput", this, c, f], b.restart)
                        }*/
                    }
                    c.j++
                }
                return null
            },</pre>

If I comment out the try/catch block, the code shows in the Chrome console this error and stops in both desktop and tablet, but it does not enter in an infinite loop as before:

<pre>
Uncaught Error: restartMathJax.js?config=TeX-AMS_HTML.js:2342 MathJax.Hub.RestartAfterMathJax.js?config=TeX-AMS_HTML.js:2238 MathJax.Hub.prepareOutputMathJax.js?config=TeX-AMS_HTML.js:211 CALLBACK.executeMathJax.js?config=TeX-AMS_HTML.js:189 cbMathJax.js?config=TeX-AMS_HTML.js:489 BASE.Object.Subclass.ProcessMathJax.js?config=TeX-AMS_HTML.js:479 BASE.Object.Subclass.PushMathJax.js?config=TeX-AMS_HTML.js:2055 MathJax.Hub.takeActionMathJax.js?config=TeX-AMS_HTML.js:2031 MathJax.Hub.ProcessMathJax.js?config=TeX-AMS_HTML.js:211 CALLBACK.executeMathJax.js?config=TeX-AMS_HTML.js:189 cbMathJax.js?config=TeX-AMS_HTML.js:489 BASE.Object.Subclass.ProcessMathJax.js?config=TeX-AMS_HTML.js:479 BASE.Object.Subclass.PushMathJax.js?config=TeX-AMS_HTML.js:2015 MathJax.Hub.TypesetMathJax.js?config=TeX-AMS_HTML.js:211 CALLBACK.executeMathJax.js?config=TeX-AMS_HTML.js:189 cbMathJax.js?config=TeX-AMS_HTML.js:489 BASE.Object.Subclass.ProcessMathJax.js?config=TeX-AMS_HTML.js:508 BASE.Object.Subclass.callMathJax.js?config=TeX-AMS_HTML.js:361 WAITEXECUTEMathJax.js?config=TeX-AMS_HTML.js:189 cbMathJax.js?config=TeX-AMS_HTML.js:211 CALLBACK.executeMathJax.js?config=TeX-AMS_HTML.js:189 cb
</pre>

**Comment by ManuelLlorens** [comment]
More info about my late night tests:

All 2.x CDN versions freezes. 
MathJax demo page freezes.
MathML output (although obviously not showing the equations in Chrome) does not freeze. 

This: \[a+b\] does work, but this: \[a_1\] does not. A page with many times this equation does work: \[a+b+c+d+e+f+g+h\]. 

The freeze seems to happen asynchronously. Sometimes it happens at 100% typesetting and the equations appear just before the freeze.

**Comment by dpvc** [comment]
@ManuelLlorens, I really appreciate your willingness to try to debug the issue, especially since we are not able to reproduce it ourselves.  Unfortunately, the error you are finding is not the problem, but part of the normal operation of MathJax.  MathJax uses errors and error-trapping to handle its asynchronous actions like loading files or waiting for fonts to arrive.

For example, when MathJax needs to load a file, it starts the download, then throws a "restart" error (that includes a pointer to the callback that will be called when the asynchronous action is complete).  The `try`/`catch` that you have removed traps that error, and when the callback is completed, it restarts the processing of that equation again.

In your case, you can see that the uncaught error is a "restart" error, and that it comes from the `RestartAfter()` function, so this error is part of MathJax's synchronization with asynchronous actions, not a "real" untapped error.  In any case, this is normal operation for MathJax, not an actual problem.

Note that the `catch` portion of the code checks for a restart error if the trapped error *isn't* a restart, it produces a message, saves the error, and goes on to the next equation.  So non-reset errors (true errors) are reported and recorded.  If you want to check for real errors, you should do it in the `catch` block that you have commented out (after the `lastPrepError` is set would be a good place).

It is unlikely that the problem you are having will be one that is trapped in one of the `try`/`catch` blocks.  If it were, MathJax would already be handling the error.  The problem will either be an error occurring *outside* of a `try`/`catch` block (where it will cause MathJax to stop running), or an error in the browser that is so bad that it can't be trapped by `try`/`catch`, or an infinite loop (which will not be something that throws an error).  The first should be reported in the console where you can see it.  The second is the hardest to deal with, and is likely the one you have here.  The infinite loop is a possibility, but usually the browser will stop with a message that javascript seems to be running too long.  So I suspect the second kind, which is the hardest to debug and understand.

My own feeling is that it is likely a memory management problem in Chrome (which certainly has some), and that there may be little we can do about it.

If you really want to try to find this yourself, I'd recommend using the unpacked version of MathJax rather than the packed one.  Just use `http://cdn.mathjax.org/mathjax/latest/unpacked/MathJax.js` instead.  That will give you proper line breaks and variable names, so it should be easier to deal with.

Again, thanks for working on this.  Let us now if we can help you further in your efforts.

**Comment by vyznev** [comment]
@dpvc: Are your *sure* it's not an infinite restart loop?  Just eyeballing the code quoted by @ManuelLlorens, the following catch block (here shown before minification) looks suspect:

        } catch (err) {
          if (!err.restart) {
            MathJax.Message.Set(["PrepError","Error preparing %1 output (%2)",id,method],null,600);
            MathJax.Hub.lastPrepError = err;
            state.j++;
          }
          return MathJax.Callback.After(["prepareOutput",this,state,method],err.restart);
        }

Note that there's no `continue;` inside the `if` block, so if this code even catches a non-restart error, it will execute the `if` block, but then *still tries to restart the action*.

(Of course, the fact that this -- if this is indeed the cause -- hangs Chrome, rather than simply giving a message about a long-running script, is presumably still a bug in Chrome.  But at least we *may* now have finally discovered what triggers it.)

Manuel, can you check if adding a `continue;` statement after `state.j++;` inside the `if` block fixes the problem for you?  (I'd try it myself, but again, I don't have the computer that exhibits this bug at hand. I may give it a try tomorrow or later this week.)

**Comment by ManuelLlorens** [comment]
Thanks a lot for the detailed info. Another try, this time I have remote debugged it (never tried this before, but works like a charm).

I have get to this, I hope it makes any sense for you, although I agree that it looks like something Chrome is doing wrong.

I have traced the execution to a point when, after the equation processing is done, this line in MathJax.js:
<pre>  //
  //  Create a callback from an associative array
  //
  var CALLBACK = function (data) {
    var cb = function () {return arguments.callee.execute.apply(arguments.callee,arguments)};</pre>

Is called when <pre>arguments.callee</pre> value is:
<pre>
function (check,font,callback) {
      if (check.time(callback)) return;
      if (HTMLCSS.Font.testFont(font)) {callback(check.STATUS.OK)}
        else {setTimeout(check,check.delay)}
    }
</pre>

<pre>arguments.calle.execute</pre> is <pre>function () {this.hook.call(this.object,this,this.data[0],this.data[1])}</pre>, <pre>this</pre> is the native <pre>window</pre> object, so <pre>this.object</pre>, <pre>this.data</pre> and <pre>this.hook</pre> are all <pre>undefined</pre>.

After execution of <pre>var cb = function () {return arguments.callee.execute.apply(arguments.callee,arguments)};</pre> Chrome freezes in every try I have done, even if I try to step into.

**Comment by ManuelLlorens** [comment]
vyznev, I am now almost sure that this is not a infinite loop but a problem in Chrome that makes it freeze.

**Comment by dpvc** [comment]
@vyznev, I'm pretty sure it's not a restart loop.

I did make an error in my previous message, which was to say that this routine was looping through equations.  It isn't, it is looping through the various output jax.  The `prepareOutput` function loops throughout the output jax that are needed (usually there is only one), and for each output jax, it calls either its `PreProcess` or `PostProcess` function, trapping errors and processing restarts.  If the output jax returns a restart, the action is restarted after the asynchronous event is completed.

Your concern about the case of a *non-restart* error is understandable, but the important point to note is the line

    state.j++

which increments the counter used to tell which output jax to process.  This means that when the restart occurs, the offending output jax will be skipped, and the error will not be repeated, unless the *next* output jax also has an error.  But since `state.j` is incremented on each error, you will eventually go through all the jax and stop, so that will not cause an infinite loop.

You are correct that it would be better not to call `MathJax.Callback.After()` in the case of a non-reset error, but it turns out not to cause a problem, since the restart will occur immediately and is effectively the same as going through the loop again directly.  It would be more efficient not to do that call, but since non-restart errors shouldn't occur anyway, and this call is made only once for each output jax (usually once), the inefficiency is minimal.  It should be fixed, though.

It is technically possible for *reset* errors to cause a loop (since they go to the same output jax a second time).  But MathJax doesn't cause a reset twice for the same condition, and so loops like that shouldn't occur.  (For example, a reset to load any given file should only happen once, since each file request is recorded, and if a request has already been made, it is not repeated.)

Also note that the error reported by @ManuelLlorens is clearly a reset error, not a *non-reset*.  Since reset errors are a normal part of MathJax operation, this would not explain the fact that he is having the problem and we can't reproduce it.  That is, in general, if he needs a file to be loaded, so would everyone else.  A loop in MathJax's use of reset errors should be reproducible by everyone.  The fact that it is only occurring for him and a few others, but not everyone, and that it happens on many different pages (not dependent on the content) suggests that it is not due to resets.  Since non-reset errors can't loop (the `state.j` will run out), I conclude that this routine is not the source of the problem.

**Comment by dpvc** [comment]
@ManuelLlorens, I will have to look more closely at your results and see if I can make sense of them.  Thanks again for your efforts.

**Comment by dpvc** [comment]
@ManuelLlorens, in your earlier message, you give the file

``` html
<html>
<head>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML.js">
   </script>
</head>
<body>
\[a_1\]
</body>
</html>
```
as an example that causes Chrome to fail for you.  

A couple of questions:

1.  Will you verify that it is this file (with the call to the CDN) that fails, and not one that uses your modified MathJax.js?

2.  Do I understand correctly that if you change `a_1` to `a+1` then you don't get the hang?

3.  @vyznev, does this file fail for you (I assume yes, since everything else does)?  Also when `a_1` is replaced by `a+1`?

Both vyznev and ManuelLlorens errors seem to be occurring during the checking for the availability of the MathJax font (vyznev's original post shows this in the loading message and the signal where the problem occurs), and ManuelLlorens' error above is occurring in the code that is checking for a font.  So I'm wondering if the problem is in Chrome's web-font code.  The font detection is pretty fragile, and I would like to rework it, so perhaps that would help.  But that isn't a short-term project.

**Comment by ManuelLlorens** [comment]
Yes, @dpvc, this file with the CDN link fails for me. 

Yes, it does not fail if I change `a_1` for `a+1`.

**Comment by pkra** [comment]
[off-topic but thanks so much @ManuelLlorens and @vyznev for your effort -- super awesome to make some progress on this elusive bug!]

**Comment by vyznev** [comment]
@dpvc: Just confirming, yes, the file in [your previous comment](https://github.com/mathjax/MathJax/issues/548#issuecomment-68658329) freezes for me. It does not freeze if I change `a_1` to `a+1`.

(Also checked that, yes, this only happens with the HTML-CSS renderer; the SVG renderer works fine for me.)

**Comment by ManuelLlorens** [comment]
In my case, more complex examples are freezing the SVG output also.

**Comment by dpvc** [comment]
Thanks, guys, that makes for something we can work with the try to find the issue.  If either of you is willing to keep trying the debugging, I can send some things to try (but it will involve changing the MathJax code, so you will need to use your own copy of MathJax rather than the CDN version.

**Comment by vyznev** [comment]
@ManuelLlorens:
Hmm, I didn't test anything *really* complex, but at least the following quick test works for me. Does it fail for you?

    <html>
    <head>
    <script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_SVG.js"></script>
    </head>
    <body>
    \begin{aligned}
    \int_0^1 \int_y^1 \int_0^y f(x,y,z) \;dz\,dx\,dy
    &= \iiint\limits_{\mathbb R^3} f(x,y,z) \,\mathbf 1_{0\le z \le y \le x \le 1} \;dz\,dx\,dy \\
    &= \iiint\limits_{\mathbb R^3} f(x,y,z) \,\mathbf 1_{0\le z \le y \le x \le 1} \;dx\,dy\,dz \\
    &= \int_0^1 \int_z^1 \int_y^1 f(x,y,z) \;dx\,dy\,dz
    \end{aligned}
    </body>
    </html>

**Comment by vyznev** [comment]
@dpvc: Sure, if you have any ideas, please do share them. I was going to try fiddling with the code anyway.

**Comment by pkra** [comment]
@dpvc I could upload a branch to the CDN.

**Comment by vyznev** [comment]
@dpvc: It gets to the end for me, before freezing. The last console output I see is:

    Hub: End Math Output,[object HTMLBodyElement],Process
    VM57:3 Hub: End Math,[object HTMLBodyElement],Process
    VM57:3 Hub: End Process,[object HTMLBodyElement]
    VM57:2 Startup: End Typeset
    VM57:2 Startup: End
    VM57:4 ***** MathJax Queue Finished *****

It's getting late here, and I'm going to get some sleep, but if I've got some time tomorrow, my plan is to install WinDBG on this laptop and see if I can take a peek inside the hung Chrome process and see what it's doing.

**Comment by ManuelLlorens** [comment]
@vyznev, your sample with SVG rendering worked for me.

**Comment by ManuelLlorens** [comment]
This is the full output in my console window. It does not show the equation `a_1` and the Loading message does not disappear:

<pre>
 Startup: Begin
VM44:5 Startup: Begin Config
VM44:5 Startup: onLoad
VM44:5 Startup: TeX noErrors Ready
VM44:5 Startup: End Config
VM44:5 Startup: mml Jax Config
VM44:5 Startup: mml Jax Require
VM44:5 Startup: mml Jax Startup
VM44:5 Startup: mml Jax Ready
VM44:5 Startup: toMathML Ready
VM44:5 Startup: TeX Jax Config
VM44:5 Startup: TeX Jax Require
VM44:5 Startup: TeX Jax Startup
VM44:5 Startup: TeX Jax Ready
VM44:5 Startup: TeX noUndefined Ready
VM44:5 Startup: TeX AMSmath Ready
VM44:5 Startup: TeX AMSsymbols Ready
VM44:5 Startup: Begin Cookie
VM44:5 Startup: End Cookie
VM44:5 Startup: Begin Styles
VM44:5 Startup: MathEvents Ready
VM44:5 Startup: MathMenu Ready
VM44:5 Startup: MathZoom Ready
VM44:5 Startup: End Styles
VM44:5 Startup: Begin Jax
VM44:5 Startup: End Jax
VM44:5 Startup: Begin Extensions
VM44:5 Startup: End Extensions
VM44:5 Startup: Begin Typeset
VM44:6 Hub: Begin PreProcess,[object HTMLBodyElement]
VM44:6 Hub: End PreProcess,[object HTMLBodyElement]
VM44:6 Hub: Begin Process,[object HTMLBodyElement]
VM44:6 Hub: Begin Math,[object HTMLBodyElement],Process
VM44:6 Hub: Begin Math Input,[object HTMLBodyElement],Process
VM44:6 Hub: End Math Input,[object HTMLBodyElement],Process
VM44:5 Startup: HTML-CSS Jax Config
VM44:5 Startup: HTML-CSS Jax Require
VM44:5 Startup: HTML-CSS Jax Startup
VM44:5 Startup: HTML-CSS Jax Ready
VM44:6 Hub: Begin Math Output,[object HTMLBodyElement],Process
VM44:5 Startup: HTML-CSS Jax - Web-Font TeX/Math/Italic
jax.js?rev=2.4-beta-2:19 Resource interpreted as Font but transferred with MIME type application/x-woff: "http://moodle.colegiomontpellier.es/mathjax/fonts/HTML-CSS/TeX/woff/MathJax_Main-Regular.woff?rev=2.4-beta-2".
jax.js?rev=2.4-beta-2:19 Resource interpreted as Font but transferred with MIME type application/x-woff: "http://moodle.colegiomontpellier.es/mathjax/fonts/HTML-CSS/TeX/woff/MathJax_Math-Italic.woff?rev=2.4-beta-2".
jax.js?rev=2.4-beta-2:19 Resource interpreted as Font but transferred with MIME type application/x-woff: "http://moodle.colegiomontpellier.es/mathjax/fonts/HTML-CSS/TeX/woff/MathJax_Size1-Regular.woff?rev=2.4-beta-2".
VM44:5 Startup: HTML-CSS Jax - Web-Font TeX/Main/Regular
VM44:6 Hub: New Math,MathJax-Element-1
VM44:6 Hub: End Math Output,[object HTMLBodyElement],Process
VM44:6 Hub: End Math,[object HTMLBodyElement],Process
VM44:6 Hub: End Process,[object HTMLBodyElement]
VM44:5 Startup: End Typeset
VM44:5 Startup: End
VM44:7 ***** MathJax Queue Finished *****
</pre>

**Comment by vyznev** [comment]
OK, some updates from [messing around with WinDBG](http://www.chromium.org/developers/how-tos/debugging#TOC-Debugging-with-WinDBG):

It appears that it's the renderer process that's getting caught in an infinite loop.  And yes, the loop seems to be inside Blink code.

I don't have a lot of experience using WinDBG, and haven't been able to do any active debugging (trying to actually run chrome.exe from within the debugger doesn't seem to work for me), but one thing that *does* work is attaching to the stuck renderer process in non-invasive mode, and taking a stack dump.  Here's a typical one:

	ChildEBP RetAddr  
	(Inline) -------- chrome_child!WTF::RawPtr<blink::Node>::operator class blink::Node * [c:\b\build\slave\win\build\src\third_party\webkit\source\wtf\rawptr.h @ 119]
	(Inline) -------- chrome_child!blink::Node::nextSibling [c:\b\build\slave\win\build\src\third_party\webkit\source\core\dom\node.h @ 189]
	0049ea48 532cbce0 chrome_child!blink::TextIterator::advance+0x30d [c:\b\build\slave\win\build\src\third_party\webkit\source\core\editing\textiterator.cpp @ 491]
	0049eae8 533b71b4 chrome_child!blink::TextIterator::rangeLength+0x4d [c:\b\build\slave\win\build\src\third_party\webkit\source\core\editing\textiterator.cpp @ 2239]
	0049eb3c 533b7433 chrome_child!blink::AXRenderObject::indexForVisiblePosition+0x16b [c:\b\build\slave\win\build\src\third_party\webkit\source\core\accessibility\axrenderobject.cpp @ 1783]
	^^^^^ THE PART ABOVE THIS LINE VARIES ^^^^^
	0049eb98 53b25ead chrome_child!blink::AXRenderObject::lineBreaks+0x101 [c:\b\build\slave\win\build\src\third_party\webkit\source\core\accessibility\axrenderobject.cpp @ 1819]
	0049ebc4 53cd6cdc chrome_child!blink::WebAXObject::lineBreaks+0x33 [c:\b\build\slave\win\build\src\third_party\webkit\source\web\webaxobject.cpp @ 909]
	0049ee54 53ca7990 chrome_child!content::BlinkAXTreeSource::SerializeNode+0xb50 [c:\b\build\slave\win\build\src\content\renderer\accessibility\blink_ax_tree_source.cc @ 340]
	0049efa8 53ca7b12 chrome_child!ui::AXTreeSerializer<blink::WebAXObject>::SerializeChangedNodes+0x218 [c:\b\build\slave\win\build\src\ui\accessibility\ax_tree_serializer.h @ 424]
	0049f0fc 53ca7cdf chrome_child!ui::AXTreeSerializer<blink::WebAXObject>::SerializeChangedNodes+0x39a [c:\b\build\slave\win\build\src\ui\accessibility\ax_tree_serializer.h @ 460]
	0049f11c 53ca7577 chrome_child!ui::AXTreeSerializer<blink::WebAXObject>::SerializeChanges+0x18b [c:\b\build\slave\win\build\src\ui\accessibility\ax_tree_serializer.h @ 330]
	0049f298 52e13147 chrome_child!content::RendererAccessibilityComplete::SendPendingAccessibilityEvents+0x159 [c:\b\build\slave\win\build\src\content\renderer\accessibility\renderer_accessibility_complete.cc @ 218]
	(Inline) -------- chrome_child!base::internal::RunnableAdapter<void (__thiscall content::ChildThread::*)(void)>::Run+0x9 [c:\b\build\slave\win\build\src\base\bind_internal.h @ 134]
	0049f2a8 52e1310b chrome_child!base::internal::InvokeHelper<1,void,base::internal::RunnableAdapter<void (__thiscall content::ChildThread::*)(void)>,void __cdecl(base::WeakPtr<content::ChildThread> const &)>::MakeItSo+0x37 [c:\b\build\slave\win\build\src\base\bind_internal.h @ 882]
	0049f2bc 52bba6c2 chrome_child!base::internal::Invoker<1,base::internal::BindState<base::internal::RunnableAdapter<void (__thiscall content::ChildThread::*)(void)>,void __cdecl(content::ChildThread *),void __cdecl(base::WeakPtr<content::ChildThread>)>,void __cdecl(content::ChildThread *)>::Run+0x15 [c:\b\build\slave\win\build\src\base\bind_internal.h @ 1169]
	(Inline) -------- chrome_child!base::Callback<void __cdecl(void)>::Run+0xb [c:\b\build\slave\win\build\src\base\callback.h @ 401]
	0049f364 52bba169 chrome_child!base::debug::TaskAnnotator::RunTask+0x32c [c:\b\build\slave\win\build\src\base\debug\task_annotator.cc @ 62]
	0049f39c 52bb9ff4 chrome_child!base::MessageLoop::RunTask+0xe4 [c:\b\build\slave\win\build\src\base\message_loop\message_loop.cc @ 449]
	(Inline) -------- chrome_child!base::MessageLoop::DeferOrRunPendingTask+0x115 [c:\b\build\slave\win\build\src\base\message_loop\message_loop.cc @ 456]
	0049f4e0 52bbc29f chrome_child!base::MessageLoop::DoWork+0x375 [c:\b\build\slave\win\build\src\base\message_loop\message_loop.cc @ 566]
	0049f504 52bb809c chrome_child!base::MessagePumpDefault::Run+0xc8 [c:\b\build\slave\win\build\src\base\message_loop\message_pump_default.cc @ 33]
	(Inline) -------- chrome_child!base::internal::PlatformThreadLocalStorage::GetTLSValue+0x7 [c:\b\build\slave\win\build\src\base\threading\thread_local_storage_win.cc @ 30]
	(Inline) -------- chrome_child!base::ThreadLocalStorage::StaticSlot::Get+0xc [c:\b\build\slave\win\build\src\base\threading\thread_local_storage.cc @ 231]
	0049f520 52bb9b3b chrome_child!tracked_objects::ThreadData::Get+0x28 [c:\b\build\slave\win\build\src\base\tracked_objects.cc @ 334]
	0049f530 52bb9a69 chrome_child!tracked_objects::TaskStopwatch::TaskStopwatch+0x3f [c:\b\build\slave\win\build\src\base\tracked_objects.cc @ 860]
	0049f55c 52bba25c chrome_child!base::RunLoop::Run+0x2d [c:\b\build\slave\win\build\src\base\run_loop.cc @ 55]
	0049f580 52c247ae chrome_child!base::MessageLoop::Run+0x46 [c:\b\build\slave\win\build\src\base\message_loop\message_loop.cc @ 309]
	0049f808 52bb2551 chrome_child!content::RendererMain+0x292 [c:\b\build\slave\win\build\src\content\renderer\renderer_main.cc @ 231]
	0049f81c 52bb24cd chrome_child!content::RunNamedProcessTypeMain+0x61 [c:\b\build\slave\win\build\src\content\app\content_main_runner.cc @ 420]
	0049f87c 52b9d251 chrome_child!content::ContentMainRunnerImpl::Run+0x66 [c:\b\build\slave\win\build\src\content\app\content_main_runner.cc @ 769]
	0049f88c 52b9c2b5 chrome_child!content::ContentMain+0x23 [c:\b\build\slave\win\build\src\content\app\content_main.cc @ 19]
	0049f8d4 00267573 chrome_child!ChromeMain+0x61 [c:\b\build\slave\win\build\src\chrome\app\chrome_main.cc @ 60]
	0049f964 00266f76 chrome!MainDllLoader::Launch+0x15f [c:\b\build\slave\win\build\src\chrome\app\client_util.cc @ 316]
	0049f9a8 002893ca chrome!wWinMain+0x5a [c:\b\build\slave\win\build\src\chrome\app\chrome_exe_main_win.cc @ 115]
	0049f9f4 763b338a chrome!__tmainCRTStartup+0xfd [f:\dd\vctools\crt\crtw32\startup\crt0.c @ 251]
	0049fa00 77749f72 KERNEL32!BaseThreadInitThunk+0xe
	0049fa40 77749f45 ntdll!__RtlUserThreadStart+0x70
	0049fa58 00000000 ntdll!_RtlUserThreadStart+0x1b

Notice the extra line I added, that says `THE PART ABOVE THIS LINE VARIES`.  Everything below it seems to be essentially constant (except for minor variation in addresses between different instances of the renderer, etc.) in all the dozen or so stack dumps I've taken so far.  This seems to at least *suggest* the possibility that the infinite loop may be happening inside `blink::AXRenderObject::lineBreaks`.

**Comment by dpvc** [comment]
Thanks for the additional information.  You guys are awesome!

I agree that the hang is occurring in Blink itself, not in MathJax.  But something MathJax is doing is certainly triggering the problem.  The question is what that is, and how we get our hands on it.

I'll have to think about what the next test should be.  I'm wondering if some nodes in the DOM have been mis-linked, as one of the calls above is `nextSibling`, so if there is a loop in the DOM structure, that might account for it.  We might be able to have MathJax walk the DOM tree after it finishes, to see if it can identify such a loop.

Alternatively, we can prevent MathJax from inserting some of its output and see if we can find which piece is causing the trouble.  I'll have to put something together for you guys to try.

**Comment by pkra** [comment]
Just a quick question: are we already at a point where we could try to informally reach out to the Blink team?

**Comment by ManuelLlorens** [comment]
Using WinDBG is out of my possibilities because I am remote debugging through Chrome for Android in a non rooted tablet. 

But I will be pleased to make any tests that could be performed with this method.

**Comment by pkra** [comment]
@dpvc ok. I've updated title and labels; reopening.

**Comment by dpvc** [comment]
Here are a couple of other things to try:  

* Do you get the same results if you use in-line math via `\(...\)` rather than `\[...\]`?  

* If you turn off font matching, does that change anything?  Use
    
        MathJax.Hub.Config({
          "HTML-CSS": {matchFontHeight: false}
        });
    
    to turn that off.

**Comment by ManuelLlorens** [comment]
No difference for me in any of the two cases.

**Comment by ManuelLlorens** [comment]
I have notice this: sometimes, when the browser freezes, the MathJax output message is: `Loading web-font TeX/Math/Italic` instead of `Typesetting Math x%`.

**Comment by nkurz** [comment]
I'm currently experiencing the same symptoms with 39.0.2171.95 (64-bit) on OSX 10.9.5.   Manuel's SVG example works fine, but the HTML-CSS version hangs.  http://mathjax.org/ loads correctly, but http://cdn.mathjax.org/mathjax/latest/test/sample-signals.html hangs.  

I haven't tried restarting or clearing cache yet.  If there is any information or logs that I can help provide, I'd be glad to try help.  I'm Linux low-level literate, but not familiar with the tools available on OSX.  I can install software as necessary.

I'll hold off on trying to fix the problem for a day or two in case there's something useful I can do.

**Comment by ManuelLlorens** [comment]
I suppose someone has already done this, but just to confirm I have tested this with MathJax 2.5 beta and it also freezes.

**Comment by H2KFL** [comment]
@nkurz could you update Chrome to 40.0.2214.91 m (latest). It works for me. I've exprienced hang up with Chrome for Windows 7 64bit. But it seems fine now.

**Comment by nkurz** [comment]
I'm currently running 39.0.2171.99 (64-bit) on OSX, and all appears
to be working correctly.  The bug persisted until I restarted Chrome.
 Attempts to clear caches and other in-browser actions had no effect.
 I'm not sure if a restart alone would have been enough to solve it,
as it updated automatically at the same time.

**Comment by dpvc** [comment]
In Chrome reseting the cache is not enough to clear MathJax (I'm not sure why).  You do have to restart the browser as well.

**Comment by ManuelLlorens** [comment]
In Android 5.0.2; Nexus 7 Build/LRX22G, Chrome 40.0.2214.89 still hungs.

**Comment by ofb** [comment]
I'm running android 5.0.2 and chrome 40.0.2214.89 as well on a nexus 10. The nexus 10 has been hanging on mathjax for what is probably over a year; I just haven't seen much discussion about it online before now, so I didn't pursue it. The maddening thing for me is that sometimes I have opened pages (usually on mathoverflow.net) that freeze 4 times in a row in different places, and then on the fifth time make it through and render, but then the tab freezes completely and you can't click links or even move the page up and down. It'd be great if this ever gets fixed.

**Comment by nitinsurana** [comment]
I'm having a similar issue, not sure if this is the as the mentioned. I'm able to reproduce it on Firefox also.
We have an html-editor (redactor to be precise), we insert MathML inside it and call
 MathJax.Hub.Queue(["Typeset", MathJax.Hub]); 
After inserting & rendering 10 mathml expressions, the editor goes slow. If I increase the number to 100 and try to move/navigate/type in other/same fields on the page, it just hangs.

I'm going to try to remove redactor (htmleditor) and just have a div with contenteditable and insert mathml and render them with  MathJax.Hub.Queue(["Typeset", MathJax.Hub]);  and see if it still hangs.

**Comment by pkra** [comment]
=> marked as abandoned.

Not looking forward to this Frankenbug resurfacing.

**Comment by adamwulf** [comment]
I was able to create a test case that reproduces the issue 100% of the time with the current master branch (commit 2965eab5).

The test case can be found here: www.wikipediacorpus.com/Chapter1.html - Opening that URL will cause MathJax to hang in an infinite loop.

The issue seems to occur when the browser fails to load a supporting resource file for some reason. In the example URL, i forced MathJax/jax/output/HTML-CSS/autoload/mtable.js to 404 by renaming it to mtable2.js on the server. I'm purposefully causing the break here, but there's tons of reasons why a browser might fail to load a resource, which explains why it's been so hard to reproduce.

When that file fails to load, the processOutput function circles in an infinite loop, continually trying to restart. I've added a console.log() to that method, which prints out the state.i value, and you can see in the JavaScript console that the value never increments above 7. The line that is throwing the restart error is line 2334:

```
        result = MathJax.OutputJax[jax.outputJax].Process(script,state);
```

As a workaround, the following seems to fix the issue:

```
        try{
                result = MathJax.OutputJax[jax.outputJax].Process(script,state);
        }catch(e){
                state.i++;
                throw e;
        }
```


Otherwise, that line throws a restart Error, having never incremented state.i, and the whole process repeats itself.

I ran into this issue while working on MacDown, a native markdown client for Mac, which uses MathJax in its preview window ([issue 807](https://github.com/MacDownApp/macdown/issues/807)). On every edit, the preview window is reloaded, which occasionally causes scripts to fail to load, which sometimes causes this infinite recursion, hanging the app.

**Comment by vyznev** [comment]
@adamwulf I suspect that may be a different issue than what's discussed above. Still, it does look like a genuine MathJax bug in any case.

My Chromium 64.0.3282.167 doesn't hang on your test page, but it does peg one CPU core at 100% for a few seconds before giving up with the following error on the console:

```
MathJax.js?config=TeX-AMS-MML_HTMLorMML:2378
Uncaught RangeError: Maximum call stack size exceeded
    at Object.processError (MathJax.js?config=TeX-AMS-MML_HTMLorMML:2378)
    at Object.processOutput (MathJax.js?config=TeX-AMS-MML_HTMLorMML:2356)
    at Function.execute (MathJax.js?config=TeX-AMS-MML_HTMLorMML:243)
    at cb (MathJax.js?config=TeX-AMS-MML_HTMLorMML:225)
    at Function.AFTER [as After] (MathJax.js?config=TeX-AMS-MML_HTMLorMML:374)
    at Object.processError (MathJax.js?config=TeX-AMS-MML_HTMLorMML:2381)
    at Object.processOutput (MathJax.js?config=TeX-AMS-MML_HTMLorMML:2356)
    at Function.execute (MathJax.js?config=TeX-AMS-MML_HTMLorMML:243)
    at cb (MathJax.js?config=TeX-AMS-MML_HTMLorMML:225)
    at Function.AFTER [as After] (MathJax.js?config=TeX-AMS-MML_HTMLorMML:374)
```

(That's the partial stack trace shown in Chromium's console by default. Clicking the triangle to expand the full stack trace just shows the same five function loop repeated over and over and over.)

**Comment by adamwulf** [comment]
I went ahead and opened a separate issue #1936 in case the cause is different than this older issue.

**Comment by dpvc** [comment]
You are right, it is a different issue.  Thanks for opening the other one.

**Comment by jingpengw** [comment]
this issue have a temporal fix:
https://github.com/jupyter/jupyter/issues/227

Using this command:
jupyter nbconvert --ClearOutputPreprocessor.enabled=True --inplace Notebook.ipynb
(just replace Notebook.ipynb with your own file name)

**Comment by ricardobeat** [comment]
Isn't it crazy that the bug still hasn't been found 6 years later? 🤕 

On the bright side I haven't seen this issue in the past half decade. Looking at the code mentioned in https://github.com/mathjax/MathJax/issues/548#issuecomment-68618832: https://github.com/mathjax/MathJax/blob/419b0a6eee7eefc0f85e47f7d4f8227ec28b8e57/unpacked/MathJax.js#L2293-L2315, I think it's clear the error is simply that `state.j` is not incremented at all in the `catch` block when `err.restart == true`, leading to infinite retries of the failed op. According to @dpvc's explanation:

> when MathJax needs to load a file, it starts the download, then throws a "restart" error (that includes a pointer to the callback that will be called when the asynchronous action is complete). The try/catch that you have removed traps that error, and when the callback is completed, it restarts the processing of that equation again

And then the use of timers somewhere in the callback system would prevent Chrome from triggering either a stack overflow, or the "script taking too long" message. This could be mitigated by keeping a counter and giving up after X retries.

**Comment by dpvc** [comment]
> I think it's clear the error is simply that `state.j` is not incremented at all in the `catch` block when `err.restart == true`, leading to infinite retries of the failed op.

While that was a suggestion at one point for what the problem was, it turns out not to be the case, as further investigation showed that MathJax completed its operations before the hang occurred.  (See [this comment](https://github.com/mathjax/MathJax/issues/548#issuecomment-68801800).)  This shows that it is not an infinite look inside MathJax (the routine you quote above is not looping), and the issue was taken to be in Chrome itself somewhere.  Since it has not been reported again since 2015 (the 2018 case was a different issue, with was dealt with in a separate issue tracker), I suspect that updates in Chrome have resolved the problem, whatever it was.

As I [mention above](https://github.com/mathjax/MathJax/issues/548#issuecomment-68657319), MathJax doesn't produce a reset twice for the same condition (that *would* be an error), so while the code you quote could loop, in practice that doesn't occur, and didn't occur for the problem being investigated in this issue tracker.

**Comment by noelantoniooo18** [comment]
I would like to raise the same issue I encounter right now:

> Uncaught RangeError: Maximum call stack size exceeded
    at Object.processError (<anonymous>:19:41875)
    at Object.processOutput (<anonymous>:19:41327)
    at Function.execute (<anonymous>:19:3242)
    at cb (<anonymous>:19:2819)
    at Function.AFTER [as After] (<anonymous>:19:6707)
    at Object.processError (<anonymous>:19:41960)
    at Object.processOutput (<anonymous>:19:41327)
    at Function.execute (<anonymous>:19:3242)
    at cb (<anonymous>:19:2819)
    at Function.AFTER [as After] (<anonymous>:19:6707)
    
    
![image](https://user-images.githubusercontent.com/90504172/186799380-9737e273-3737-4c1e-bc53-6517886f8788.png)

Can anyone provide an update to this? Thank you very much.

**Comment by dpvc** [comment]
@redooo666, can you provide a link to a page where this is occurring for you?  Does it always occur, or is it intermittent?  What browser and os are you using, and what versions?

After the error occurs, if you enter `MathJax.Message.Log()` in the browser console, are there any messages about filing to load a file?

**Comment by pkra** [comment]
> if anyone needs an example of mathjax locking up the browser: https://docs.opencv.org/4.x/d2/de8/group__core__array.html

That page renders fine for me (Chrome 105.0.5195.102), even throttling down CPU 6x, switching MathJax renderers etc. (FWIW, that page uses 2.7.0 which is 6 years old now.)
